xls/dslx/stdlib/apfloat.x

```diff
@@ -183,6 +183,183 @@ pub fn unflatten<EXP_SZ:u32, SFD_SZ:u32,
   }
 }
 
+pub fn cast_from_fixed<EXP_SZ:u32, SFD_SZ:u32, UEXP_SZ:u32 = EXP_SZ + u32:1,
+  NUM_SRC_BITS:u32, EXTENEDED_SFD_SZ:u32 = SFD_SZ + NUM_SRC_BITS>
+  (to_cast: sN[NUM_SRC_BITS]) -> APFloat<EXP_SZ, SFD_SZ> {
+  // Determine sign.
+  let sign = to_cast[NUM_SRC_BITS-u32:1 : NUM_SRC_BITS];
+
+  // Determine exponent.
+  let abs_magnitude = (to_cast if sign == u1:0 else -to_cast) as uN[NUM_SRC_BITS];
+  let lz = clz(abs_magnitude);
+  let num_trailing_nonzeros = NUM_SRC_BITS as uN[NUM_SRC_BITS] - lz;
+  let exp = num_trailing_nonzeros as uN[UEXP_SZ] - uN[UEXP_SZ]:1;
+  let max_exp_exclusive = uN[UEXP_SZ]:1 << EXP_SZ as uN[UEXP_SZ] - uN[UEXP_SZ]:1;
+  let is_inf = exp >= max_exp_exclusive;
+  let bexp = bias<EXP_SZ, SFD_SZ>(exp);
+
+  // Determine significand (pre-rounding).
+  let extended_sfd = abs_magnitude ++ uN[SFD_SZ]:0;
+  let sfd = extended_sfd >>
+    (num_trailing_nonzeros - uN[NUM_SRC_BITS]:1) as uN[EXTENEDED_SFD_SZ];
+  let sfd = sfd[0 : SFD_SZ];
+
+  // Round significand (round to nearest, half to even).
+  let lsb_idx:uN[EXTENEDED_SFD_SZ] = num_trailing_nonzeros as uN[EXTENEDED_SFD_SZ]
+    - uN[EXTENEDED_SFD_SZ]:1;
+  let halfway_idx:uN[EXTENEDED_SFD_SZ] = lsb_idx - uN[EXTENEDED_SFD_SZ]:1;
```

Comment:
These type annotations shouldn't be needed...but they don't _hurt_ anything either, so...meh.

Reply:
Removed.